src/data_write.py

```diff
@@ -1236,161 +939,143 @@ def write_raw_rf_params(param):
-            param['intf_antenna_count'] = np.uint32(self.options.intf_antenna_count)
+            slice_data.data = np.concatenate(samples_list)
+            slice_data.rx_sample_rate = np.float32(data_parsing.rx_rate)
+            slice_data.num_samps = np.uint32(len(samples_list[0]) / total_ants)
+            slice_data.data_descriptors = ["num_sequences", "num_antennas", "num_samps"]
+            slice_data.data_dimensions = np.array([slice_data.num_sequences, total_ants,
+                                                   slice_data.num_samps], dtype=np.uint32)
+            slice_data.main_antenna_count = np.uint32(self.options.main_antenna_count)
+            slice_data.intf_antenna_count = np.uint32(self.options.intf_antenna_count)
 
-            for field in list(param.keys()):
-                if field not in needed_fields:
-                    param.pop(field, None)
+            write_file(output_file, slice_data, self.raw_rf_two_hr_name, "rawrf")
 
-            write_file(output_file, param, self.raw_rf_two_hr_name)
-
-            # Can only close mapped memory after its been written to disk.
-            for shm in shms:
-                shm.close()
-                shm.unlink()
+            # Can only close mapped memory after it's been written to disk.
+            for shared_mem in shared_memory_locations:
+                shared_mem.close()
+                shared_mem.unlink()
 
         def write_tx_data():
             """
             Writes out the tx samples and metadata for debugging purposes.
             Does not use same parameters of other writes.
             """
-            tx_data = None
-            for meta in aveperiod_meta.sequences:
-                if meta.tx_data is not None:
-                    tx_data = copy.deepcopy(TX_TEMPLATE)
-                    break
-
-            if tx_data is not None:
-                for meta in aveperiod_meta.sequences:
-                    meta_data = meta.tx_data
-                    tx_data['tx_rate'].append(meta_data.tx_rate)
-                    tx_data['tx_center_freq'].append(meta_data.tx_ctr_freq)
-                    tx_data['pulse_timing_us'].append(meta_data.pulse_timing_us)
-                    tx_data['pulse_sample_start'].append(meta_data.pulse_sample_start)
-                    tx_data['dm_rate'].append(meta_data.dm_rate)
-                    tx_data['tx_samples'].append(meta_data.tx_samples)
-                    tx_data['decimated_tx_samples'].append(meta_data.decimated_tx_samples)
-
-                name = dataset_name.replace('{sliceid}.', '').format(dformat='txdata')
-                output_file = dataset_location.format(name=name)
+            tx_data = SliceData()
+            for f in SliceData.type_fields('txdata'):
+                setattr(tx_data, f, [])     # initialize to a list for all fields
+
+            has_tx_data = [sqn.tx_data is not None for sqn in aveperiod_meta.sequences]
```

Comment:
This line and line 980 are not needed. You are looping through aveperiod_meta.sequences twice and checking for data twice. If sqn.tx_data is not None then do the work, else continue.

Reply:
The logic here is to write txdata to file if **any** of the slices have txdata. If we only loop through once, we won't know when we visit the first `aveperiod_meta.sequence` if the last `aveperiod_meta.sequence` contains txdata. Also, there are only ever going to be a handful of items in `aveperiod_meta.sequences`, so I think the list comprehension has a negligible impact on performance.